internal/pkg/addon/package.go

```diff
@@ -273,8 +281,129 @@ type s3Object struct {
 }
 
 func (a *Addons) uploadAddonAsset(assetPath string, forceZip bool) (s3Object, error) {
+	// make path absolute from wsPath
+	if !path.IsAbs(assetPath) {
+		assetPath = path.Join(a.wsPath, assetPath)
+	}
+
+	info, err := a.fs.Stat(assetPath)
+	if err != nil {
+		return s3Object{}, err
+	}
+
+	var asset asset
+	if forceZip || info.IsDir() {
+		asset, err = a.zipAsset(assetPath)
+	} else {
+		asset, err = a.fileAsset(assetPath)
+	}
+	if err != nil {
+		return s3Object{}, fmt.Errorf("create asset: %w", err)
+	}
+
+	s3Path := artifactpath.AddonAsset(a.wlName, asset.hash)
+
+	url, err := a.uploader.Upload(a.bucket, s3Path, asset.data)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("upload %s to s3 bucket %s: %w", assetPath, a.bucket, err)
+	}
+
+	bucket, key, err := s3.ParseURL(url)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("parse s3 url: %w", err)
+	}
+
+	fmt.Printf("Uploaded %s to s3 at: %s\n", assetPath, s3.Location(bucket, key))
 	return s3Object{
-		Bucket: a.bucket,
-		Key:    "TODO",
+		Bucket: bucket,
+		Key:    key,
+	}, nil
+}
+
+type asset struct {
+	data io.Reader
+	hash string
+}
+
+// zipAsset creates an asset from the directory or file specified by root
+// where the data is the compressed zip archive, and the hash is
+// a hash of each files name, permission, and content.
+func (a *Addons) zipAsset(root string) (asset, error) {
+	buf := &bytes.Buffer{}
+	z := zip.NewWriter(buf)
+	defer z.Close()
+
+	hash := sha256.New()
+
+	if err := a.fs.Walk(root, func(path string, info fs.FileInfo, err error) error {
+		switch {
+		case err != nil:
+			return err
+		case info.IsDir():
+			return nil
```

Comment:
just to make although we return `nil`, `Walk` will still traverse down the directory right?

Reply:
correct! you have to return [`fs.SkipDir`](https://pkg.go.dev/io/fs#SkipDir) to skip the directory but continue traversing (any other error stops entirely)